### PR TITLE
Give more details about errors while parsing want/got values

### PR DESCRIPTION
**Pull request by fadi-alkatut**
Add more info in error message

### CODE REVIEW

**Comment by pavelnikolov** [comment]
Just add a unit test and I'd be happy to merge this.

File: gqltesting/testing.go [hunk 1]
```diff
@@ -46,13 +46,14 @@ func RunTest(t *testing.T, test *Test) {
 	// Verify JSON to avoid red herring errors.
 	got, err := formatJSON(result.Data)
 	if err != nil {
-		t.Fatalf("got: invalid JSON: %s", err)
+		response := string(got[:])
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: This line is not needed, please delete it.

File: gqltesting/testing.go [hunk 2]
```diff
@@ -46,13 +46,14 @@ func RunTest(t *testing.T, test *Test) {
...
+		t.Fatalf("got: invalid JSON: %s\n response value is: %s", err, response)
 	}
 	want, err := formatJSON([]byte(test.ExpectedResult))
 	if err != nil {
-		t.Fatalf("want: invalid JSON: %s", err)
+		expected := string(want[:])
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: This line is not needed, please delete it.

File: gqltesting/testing.go [hunk 3]
```diff
@@ -46,13 +46,14 @@ func RunTest(t *testing.T, test *Test) {
 	// Verify JSON to avoid red herring errors.
 	got, err := formatJSON(result.Data)
 	if err != nil {
-		t.Fatalf("got: invalid JSON: %s", err)
+		response := string(got[:])
+		t.Fatalf("got: invalid JSON: %s\n response value is: %s", err, response)
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: Please, rewrite this as:
```Go
t.Fatalf("got: invalid JSON: %s\n response value is: %s", err, got)
```

File: gqltesting/testing.go [hunk 4]
```diff
@@ -46,13 +46,14 @@ func RunTest(t *testing.T, test *Test) {
...
 	}
 	want, err := formatJSON([]byte(test.ExpectedResult))
 	if err != nil {
-		t.Fatalf("want: invalid JSON: %s", err)
+		expected := string(want[:])
+		t.Fatalf("want: invalid JSON: %s\n expected value is: %s", err, expected)
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: Please, rewrite this as:
```Go
t.Fatalf("want: invalid JSON: %s\n expected value is: %s", err, want)
```

**Comment by fadi-alkatut** [comment]
Having more experimenting done on this issue I found the best solution is
- check for errors before expected result;
- print more details about unexpected errors;
- and only check for result if it's expected

File: gqltesting/testing.go [hunk 5]
```diff
@@ -46,13 +46,11 @@ func RunTest(t *testing.T, test *Test) {
...
 	got, err := formatJSON(result.Data)
 	if err != nil {
-		response := string(got[:])
-		t.Fatalf("got: invalid JSON: %s\n response value is: %s", err, response)
+		t.Fatalf("got: invalid JSON: %s\n response value is: [%s]", err, got)
 	}
 	want, err := formatJSON([]byte(test.ExpectedResult))
 	if err != nil {
-		expected := string(want[:])
-		t.Fatalf("want: invalid JSON: %s\n expected value is: %s", err, expected)
```
[the remark this answers is not in the document]
Comment (fadi-alkatut) [reply gqltesting/testing.go]: give more details on errors, but it wasn't very efficient

File: gqltesting/testing.go [hunk 6]
```diff
@@ -43,22 +43,22 @@ func RunTest(t *testing.T, test *Test) {
...
-		t.Fatalf("want: invalid JSON: %s", err)
-	}
-
 	checkErrors(t, test.ExpectedErrors, result.Errors)
-
-	if !bytes.Equal(got, want) {
-		t.Logf("got:  %s", got)
-		t.Logf("want: %s", want)
-		t.Fail()
+	if test.ExpectedResult != "" {
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: Can we ever expect the result to be empty string? If yes, can you, please, reverse this condition:
```go
if test.ExpectedResult == "" {
    // handle this case
}
// unindent the rest of the code
```

File: graphql_test.go [hunk 7]
```diff
@@ -247,57 +246,20 @@ func TestBasic(t *testing.T) {
...
-
-func (r *testNilInterfaceResolver) B() (interface{ Z() int32 }, error) {
-	return nil, errors.New("x")
-}
-
-func (r *testNilInterfaceResolver) C() (interface{ Z() int32 }, error) {
-	return nil, nil
-}
-
-func TestNilInterface(t *testing.T) {
```
Comment (pavelnikolov) [inline graphql_test.go]: Why is this test deleted? Do you need to rebase your PR?

File: gqltesting/testing.go [hunk 8]
```diff
@@ -43,22 +43,23 @@ func RunTest(t *testing.T, test *Test) {
...
-
-	checkErrors(t, test.ExpectedErrors, result.Errors)
-
-	if !bytes.Equal(got, want) {
-		t.Logf("got:  %s", got)
-		t.Logf("want: %s", want)
-		t.Fail()
+	if result.Errors != nil {
+		checkErrors(t, test.ExpectedErrors, result.Errors)
+	} else {
```
[the remark this answers is not in the document]
Comment (pavelnikolov) [reply gqltesting/testing.go]: For example:
```go
if result.Errors != nil && checkErrors(t, test.ExpectedErrors, result.Errors) {
    return
}
```
or you can even move the `results.Errors != nil` check inside of the `checkErrors` func. And if it returns `true` you can exit early, no? WDYT?

File: gqltesting/testing.go [hunk 9]
```diff
@@ -43,17 +43,14 @@ func RunTest(t *testing.T, test *Test) {
...
+	checkErrors(t, test.ExpectedErrors, result.Errors)
 	// Verify JSON to avoid red herring errors.
 	got, err := formatJSON(result.Data)
-	if err != nil {
+	if err != nil && test.ExpectedResult != "" {
 		t.Fatalf("got: invalid JSON: %s\n response value is: [%s]", err, got)
 	}
 	want, err := formatJSON([]byte(test.ExpectedResult))
-	if err != nil {
+	if err != nil && test.ExpectedResult != ""  {
```
Comment (pavelnikolov) [inline gqltesting/testing.go]: Why is this required? And why do you skip it if the expected result is `""`?

Comment (pavelnikolov) [reply gqltesting/testing.go]: After a second look, I understood what's going on. In case of an error, fail only if there is expected result.

Comment (fadi-alkatut) [reply gqltesting/testing.go]: Yes, because if an error happens and returns empty response it will give error when doing  `formatJSON` and this will fail the test even if you are expecting the error and an the empty result.

**Comment by pavelnikolov** [comment]
@fadi-alkatut this PR now has conflicts. Do you think it's worth resolving these? Or should the PR be closed?

**Comment by pavelnikolov** [comment]
@fadi-alkatut ping

**Comment by fadi-alkatut** [comment]
I think we went with another PR that had a cleaner solution for the problem. So this PR isn’t relevant anymore.
